### ISSUE TITLE
Results are not updated when underlying subscription changes in Flexible Sync.

### ISSUE THREAD

**Issue by kachalo** [realm/realm-swift#7859]
### How frequently does the bug occur?

All the time

### Description

I have the following setup:
```
// I define Result set for all my objects
// for the realm containing 3 objects of MyObject.
let allMyObjects : Results<MyObject> = realm.objects(MyObject.self)
Swift.print("Objects count: \(allMyObjects.count)") // Prints 3 
```

When I try to modify the subscription that have introduce or remove new objects from realm, the `Results` doesn't reflect that. However, when I open realm file in a Realm studio, it contains all objects as expected.

```
// When I modify the subscription to add new object(s)
_refreshSubscriptions(for: ["id1", "id2"]) // Still prints 3

// Or when I modify the subscription to remove
// some objects that were initially a part of a subscription
_refreshSubscriptions(for: ["id1"]) // Still prints 3
```

It is important to highlight that objects for `id1` and `id2` are available in a database and, as I mentioned before, they are in realm file.

Here's the code for `_refreshSubscriptions` method.

```
private func _refreshSubscriptions(for newIDs: Set<String>) async
{
    let realmSubscriptions = realm.subscriptions
    try! await realmSubscriptions.update
    {        
        if let currentSubscription = realmSubscriptions.first(named: "all-my-objects")
        {
            currentSubscription.updateQuery(toType: MyObject.self)
            {
                $0._id.in(newIDs)
            }
        }
        else if !newIDs.isEmpty
        {
            realmSubscriptions.append(QuerySubscription<MyObject>(name: "all-my-objects")
            {
                $0._id.in(newIDs)
            })
        }
    }
}
```

### Can you reproduce the bug?

Yes, always

### Version

v10.28.1

### What SDK flavour are you using?

MongoDB Realm (i.e. Sync, auth, functions)

### Are you using encryption?

No, not using encryption

### Platform OS and version(s)

iOS 15.5

### Build environment

Xcode version: 13.4.1 (13F100)

**Comment by reveelapp** [comment]
@kachalo,

I encountered something similar...

Have you tried checking the '`SyncSubscriptionState`' after awaiting '`_refreshSubscriptions()`' completion at call-site? - maybe test that, by returning the '`SyncSubscriptionState`' from your '`_refreshSubscriptions()`' (or check at call-site). IDK if this is the same issue I noticed, because I ended up managing states on every write/update since I noticed some quarky stuff during testing. Unfortunately, at this time we currently cannot observe the '`SyncSubscriptionState`' in Swift SDK, so that is what I do.

If it is not that, then I am curious to see what is going on...

**Comment by dianaafanador3** [comment]
Hi @kachalo are you sure this is not working. I tested the following in v10.28.1, and it is working correctly 
```
@MainActor
    func test() async throws {
        var ids: [ObjectId] = []
        try await populateFlexibleSyncData { realm in
            for i in 1...20 {
                let person = SwiftPerson(firstName: "\(#function)",
                                         lastName: "lastname_\(i)",
                                         age: i)
                realm.add(person)
                if i == 1 || i == 2 {
                    ids.append(person._id)
                }
            }
        }
        
        let user = try await logInUser(for: basicCredentials(app: self.flexibleSyncApp), app: self.flexibleSyncApp)
        var config = user.flexibleSyncConfiguration(initialSubscriptions: { subscriptions in
            subscriptions.append(QuerySubscription<SwiftPerson>(name: "all-my-objects"))
        })
        
        if config.objectTypes == nil {
            config.objectTypes = [SwiftTypesSyncObject.self, SwiftPerson.self]
        }
        let realm = try await Realm(configuration: config, downloadBeforeOpen: .once)
        let allMyObjects: Results<SwiftPerson> = realm.objects(SwiftPerson.self)
        XCTAssertEqual(allMyObjects.count, 20)
        
        
        let subs = realm.subscriptions
        try await subs.update {
            if let currentSubscription = subs.first(named: "all-my-objects") {
                currentSubscription.updateQuery(toType: SwiftPerson.self) { $0._id.in(ids) }
            } else {
                subs.append(QuerySubscription<SwiftPerson>(name: "all-my-objects") {
                    $0._id.in(ids)
                })
            }
        }
        
        XCTAssertEqual(allMyObjects.count, 2)
    }
```
Are you sure the subscription is been updated?, Also you do have to await for the update be complete before trying to read the realm again? Also, can you check if the realm is the same as the one you are getting the results?

**Comment by kachalo** [comment]
Hello @dianaafanador3,

I've spent the last three days exploring, testing, and analyzing the results. It is weird, but the issue I was having did not reproduce since I created this report.
My strategy was to run our app through testing with extensive logging, so I'll be able to spot the issue. However, it didn't reproduce though.
I have added the diagnostics log suggested by @reveelapp, showing that the subscription is indeed in the expected state.
Whatever it was, I assume it was my configuration or improper error handling on my side, not the bug in the framework.

Thank you for the valuable tips and help in resolving this problem.
